### ISSUE TITLE
bow_topk_prob are not always less than 1

### ISSUE THREAD

**Issue by nathias2004** [FranxYao/dgm_latent_bow#10]
Hi,
I see that while you are calculating the bow_topk_prob in bow_seq2seq.py lin no. 146, you are just summing up the probabilities along with the dimension 1 , but this summing up will lead to probabilities >1 and when you do log x where x>1 log x will be positive and nll will become negative.

bow_topk_prob += tf.reduce_sum(bow_prob, 1) shuld be bow_topk_prob += tf.reduce_mean(bow_prob, 1)

correct me if i am wrong

**Comment by FranxYao** [comment]
Oh thank you for pointing this out! And yes they sum up larger than one.
But considering the padding of the source sentences, instead of using:
bow_topk_prob += tf.reduce_mean(bow_prob, 1)
You might want to use
bow_topk_prob += tf.reduce_sum(bow_prob, 1) / tf.expand_dims(enc_lens, [2])
# [B, T, V] -> [B, V] / [B, 1]
I think this should work.

And additionally, note that the tf.expand_dims(enc_lens, [2])term only
serves as a scaling factor of the bow_topk_prob, even without this,
the calculation of the BOW loss and the decoder inputs should still be
correct.

Hope this helps.

Yao

**Comment by nathias2004** [comment]
hi.. I appreciate your response.
i agree that tf.expand_dims term is a scaling factor. If tf.expand_dims term is not used  the probabilities in may become > 1 and in the nll loss calculation i.e line 92 in bow_seqtoseq.py enc_loss = - enc_targets * tf.log(bow_topk_prob + 1e-6) it is possible that enc_loss may come negative. You can see and check this mathematically

**Comment by FranxYao** [comment]
Yep the enc_loss may come negative, what I mean is that even it becomes
negative, it does not change the direction of optimization, see:
without scaling:
max log p_i
with scaling:
max log C*p_i = log p_i + log C <=> max log p_i
The two are equivalent in terms of optimization.

I'm not trying to find an excuse for being wrong :) and I thank you for
pointing out my mistake! Just want to make our derivation formal.
Actually this is one of a bad habit when I am writing codes. Typically
summation/ averaging does not change the optimization direction so I was
not careful enough. Will improve!

Yao

**Comment by nathias2004** [comment]
Gotcha Yao...Thanks for the response and I really appreciate your help
